Source/OCMock/OCMBoxedReturnValueProvider.m

```diff
@@ -43,16 +58,37 @@ - (void)handleInvocation:(NSInvocation *)anInvocation
     }
 }
 
-
-- (BOOL)isMethodReturnType:(const char *)returnType compatibleWithValueType:(const char *)valueType
+- (BOOL)isMethodReturnType:(const char *)returnType compatibleWithValueType:(const char *)valueType value:(const char*)value valueSize:(size_t)valueSize
 {
     /* Same types are obviously compatible */
     if(strcmp(returnType, valueType) == 0)
         return YES;
 
-    /* Allow void* for methods that return id, mainly to be able to handle nil */
-    if(strcmp(returnType, @encode(id)) == 0 && strcmp(valueType, @encode(void *)) == 0)
-        return YES;
+    // Special casing for nil.
+    if(strcmp(returnType, @encode(id)) == 0)
+    {
+        // Check to verify that the value is actually zero.
+        if(IsZeroBuffer(value, valueSize))
+        {
+            // nil and Nil get potentially different encodings depending on the compilation
```

Comment:
If you are checking for Nil as well, couldn't that also be @encode(Class) ?

Reply:
Yes. Good call. Added.